src/main/java/org/springframework/data/relational/core/conversion/BasicRelationalConverter.java

```diff
@@ -0,0 +1,256 @@
+	 */
+	@Override
+	public <T> T createInstance(PersistentEntity<T, RelationalPersistentProperty> entity,
+			ParameterValueProvider<RelationalPersistentProperty> parameterValueProvider) {
+
+		return entityInstantiators.getInstantiatorFor(entity) //
+				.createInstance(entity, new ConvertingParameterValueProvider<>(parameterValueProvider));
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.relational.core.conversion.RelationalConverter#readValue(java.lang.Object, org.springframework.data.util.TypeInformation)
+	 */
+	@Override
+	@Nullable
+	public Object readValue(@Nullable Object value, TypeInformation<?> type) {
+
+		if (null == value) {
+			return null;
+		}
+
+		if (conversions.hasCustomReadTarget(value.getClass(), type.getType())) {
+			return conversionService.convert(value, type.getType());
+		} else {
+			return getPotentiallyConvertedSimpleRead(value, type.getType());
+		}
+	}
+
+	/*
+	 * (non-Javadoc)
+	 * @see org.springframework.data.relational.core.conversion.RelationalConverter#writeValue(java.lang.Object, org.springframework.data.util.TypeInformation)
+	 */
+	@Override
+	@Nullable
+	public Object writeValue(@Nullable Object value, TypeInformation<?> type) {
+
+		if (value == null) {
+			return null;
+		}
+
+		Class<?> rawType = type.getType();
+		RelationalPersistentEntity<?> persistentEntity = context.getPersistentEntity(value.getClass());
+
+		if (persistentEntity != null) {
+
+			Object id = persistentEntity.getIdentifierAccessor(value).getIdentifier();
+			return writeValue(id, type);
+		}
+
+		if (rawType.isInstance(value)) {
+			return getPotentiallyConvertedSimpleWrite(value);
+		}
+
+		return conversionService.convert(value, rawType);
+	}
+
+	/**
+	 * Checks whether we have a custom conversion registered for the given value into an arbitrary simple JDBC type.
+	 * Returns the converted value if so. If not, we perform special enum handling or simply return the value as is.
+	 *
+	 * @param value
+	 * @return
+	 */
+	private Object getPotentiallyConvertedSimpleWrite(Object value) {
+
+		Optional<Class<?>> customTarget = conversions.getCustomWriteTarget(value.getClass());
+
+		if (customTarget.isPresent()) {
+			return conversionService.convert(value, customTarget.get());
+		}
+
+		return Enum.class.isAssignableFrom(value.getClass()) ? ((Enum<?>) value).name() : value;
+	}
+
+	/**
+	 * Checks whether we have a custom conversion for the given simple object. Converts the given value if so, applies
+	 * {@link Enum} handling or returns the value as is.
+	 *
+	 * @param value
+	 * @param target must not be {@literal null}.
+	 * @return
+	 */
+	@Nullable
+	@SuppressWarnings({ "rawtypes", "unchecked" })
+	private Object getPotentiallyConvertedSimpleRead(@Nullable Object value, @Nullable Class<?> target) {
+
+		if (value == null || target == null || ClassUtils.isAssignableValue(target, value)) {
```

Comment:
Is it correct to simply return the `value` if it is assignable to the `target`, without checking for a `CustomReadTarget`?

I'm not sure I understand what CustomReadTargets actually are.